### PR TITLE
Flowletmanageropt

### PR DESCRIPTION
**Pull request by reshadi**
An optimization for make FlowletManager's stack a lot faster.

### CODE REVIEW

File: packages/hyperion-util/src/TimedTrigger.ts [hunk 1]
```diff
@@ -6,6 +6,9 @@
 
 type TimeoutID = number | any /* React expects Timeout?! */;
 
+
+const SUPPOORTS_IDLE_CALLBACK = typeof requestIdleCallback === 'function' && typeof cancelIdleCallback === "function";
```
Comment (schwartzmx) [inline packages/hyperion-util/src/TimedTrigger.ts:10]: minor `SUPPORTS_IDLE_CALLBACK`

File: packages/hyperion-flowlet/test/FlowletManager.test.ts [hunk 2]
```diff
@@ -31,7 +33,7 @@ describe("test FlowletManager", () => {
     expect(pops).toStrictEqual([p1, f1_1])
   });
 
-  test("test FlowletManager methods and events for async code", async () => {
+  false && test("test FlowletManager methods and events for async code", async () => {
```
Comment (schwartzmx) [inline packages/hyperion-flowlet/test/FlowletManager.test.ts:36]: should still be here?

File: packages/hyperion-flowlet/src/FlowletManager.ts [hunk 3]
```diff
@@ -6,15 +6,60 @@ import { Hook } from "@hyperion/hook";
...
+      CLEANUP_PERIOD,
+      true // use idle callback when possible. 
+    );
+  }
+
+  public cleanup(fullCleanup: boolean) {
+    if (fullCleanup) {
+      // We are in the idle period and should have nothing on the stack
+      if (this.stackSize() > 0) {
+        this._flowletStack = [];
```
Comment (schwartzmx) [inline packages/hyperion-flowlet/src/FlowletManager.ts:54]: Should we log when this happens so we know there is cases where we completely empty the stack?

File: packages/hyperion-flowlet/src/FlowletManager.ts [hunk 4]
```diff
@@ -28,14 +73,15 @@ export class FlowletManager<T extends Flowlet = Flowlet> {
...
   push(flowlet: T, forkReason?: string): T {
     const newFlowlet = forkReason && this.flowletCtor ? new this.flowletCtor(forkReason, flowlet) : flowlet;
-    this.onPush.call(flowlet, forkReason);
+    this.onPush.call(flowlet, forkReason, newFlowlet);
     this._flowletStack.push(newFlowlet);
     this.updateTop();
     return newFlowlet;
   }
-  readonly onPush = new Hook<(flowlet: T, reason?: string) => void>();
+  readonly onPush = new Hook<(flowlet: T, reason?: string, replacementtFlowlet?: T) => void>();
```
Comment (schwartzmx) [inline packages/hyperion-flowlet/src/FlowletManager.ts:84]: minor `replacementtFlowlet` to `replacementFlowlet`

**Review by schwartzmx — APPROVED** [review]
overall looks good but there is a test error, maybe related to timedtrigger / jest timers?

File: packages/hyperion-flowlet/src/FlowletManager.ts (hunk 3 above)
Comment (reshadi) [reply packages/hyperion-flowlet/src/FlowletManager.ts:54]: I have to expose some form of logging api to do this. I'll add that.

**Pull request merged**